clients/drcachesim/tracer/raw2trace.h

```diff
@@ -672,6 +679,103 @@ struct trace_header_t {
             return msg;    \
     } while (0)
 
+template <typename T> class bitset_hash_table_t {
+private:
+    const static size_t BLOCK_SIZE_BIT = 13;
+    const static size_t BLOCK_SIZE = (1 << BLOCK_SIZE_BIT);
+    const size_t BASIC_BUCKET_COUNT = (1 << 15);
+    std::unordered_map<T, std::bitset<BLOCK_SIZE>> page_table;
+    typename std::unordered_map<T, std::bitset<BLOCK_SIZE>>::iterator last_block =
+        page_table.end();
+
+    inline std::pair<T, ushort>
+    convert(T pc)
+    {
+        return std::pair<T, ushort>(
+            reinterpret_cast<T>(reinterpret_cast<size_t>(pc) & (~(BLOCK_SIZE - 1))),
+            static_cast<ushort>(reinterpret_cast<size_t>(pc) & (BLOCK_SIZE - 1) &
+                                0xFFFF));
+    }
+
+public:
+    bitset_hash_table_t()
+    {
+        static_assert(std::is_pointer<T>::value || std::is_integral<T>::value,
+                      "Pointer or integral type required");
+        page_table.reserve(BASIC_BUCKET_COUNT);
+        page_table.emplace(T(0), std::move(std::bitset<BLOCK_SIZE>()));
+        last_block = page_table.begin();
+    }
+
+    void
+    insert(T pc)
+    {
+        auto block = convert(pc);
+        if (last_block->first != block.first)
+            last_block = page_table.find(block.first);
+        auto val = block.second;
+        last_block->second[block.second] = 1;
+    }
+
+    bool
+    find_and_insert(T pc)
+    {
+        auto block = convert(pc);
+        if (block.first != last_block->first) {
+            last_block = page_table.find(block.first);
+            if (last_block == page_table.end()) {
+                last_block = (page_table.emplace(std::make_pair(
+                                  block.first, std::move(std::bitset<BLOCK_SIZE>()))))
+                                 .first;
+                last_block->second[block.second] = 1;
+                return true;
+            }
+        }
+        if (last_block->second[block.second] == 1)
+            return false;
+        last_block->second[block.second] = 1;
+        return true;
+    }
+
+    void
+    erase(T pc)
+    {
+        auto block = convert(pc);
+        if (last_block->first != block.first)
+            last_block = page_table.find(block.first);
+        last_block->second[block.second] = 0;
+    }
+
+    bool
+    find(T pc)
+    {
+        auto block = convert(pc);
+        if (block.first != last_block->first) {
+            last_block = page_table.find(block.first);
+            if (last_block == page_table.end()) {
+                last_block = (page_table.emplace(std::make_pair(
+                                  block.first, std::move(std::bitset<BLOCK_SIZE>()))))
+                                 .first;
+            }
+        }
+        return (last_block != page_table.end() && last_block->second[block.second]);
```

Comment:
== 1

Reply:
remove this function. it was not needed at all